src/utils.ts

```diff
@@ -45,9 +45,54 @@ export function coalesce<T>(valueOpt: T | null, defaultValue: T): T {
   }
 }
 
-export function toPriceInUnits(price: BigInt, decimalsBaseOpt: BigInt | null, decimalsQuoteOpt: BigInt | null): BigInt {
+export function toWei(amount: BigInt, decimalsOpt: BigInt | null): BigInt {
+  let decimals = coalesce<BigInt>(decimalsOpt, DEFAULT_DECIMALS).toI32()
+  return amount.times(TEN.pow(u8(decimals)))
+}
+
+export function greatestCommonDenominator(numerator: BigInt, denominator: BigInt): BigInt {
+  if (denominator.isZero()) {
+    return numerator
+  } else {
+    return greatestCommonDenominator(denominator, numerator.mod(denominator))
+  }
+}
+
+export function reduce(numerator: BigInt, denominator: BigInt): BigInt[] {
+  let greatestCommonDenominator = greatestCommonDenominator(numerator, denominator)
+  return [numerator.div(greatestCommonDenominator), denominator.div(greatestCommonDenominator)]
+}
+
+export function calculatePrice(
+  priceNumerator: BigInt, // 1,001997154410.220200000000000000    --- 1001997154410220200000000000000
+  priceDenominator: BigInt, // 1                                    --- 1000000000000000000
+  decimalsBaseOpt: BigInt | null, // 6
+  decimalsQuoteOpt: BigInt | null, // 18
+): BigInt[] {
   let decimalsBase = coalesce<BigInt>(decimalsBaseOpt, DEFAULT_DECIMALS).toI32()
   let decimalsQuote = coalesce<BigInt>(decimalsQuoteOpt, DEFAULT_DECIMALS).toI32()
+  log.info('[utils:calculatePrice] {} / {} with decimals {} and {}', [
+    priceNumerator.toString(),
+    priceDenominator.toString(),
+    BigInt.fromI32(decimalsBase).toString(),
+    BigInt.fromI32(decimalsQuote).toString(),
+  ])
+
+  let newNumerator: BigInt
+  let newDenominator: BigInt
+  if (decimalsBase == decimalsQuote) {
+    // Strictly not needed to handle this case separately, but saves a bit of computation to the indexer
+    newNumerator = priceNumerator
+    newDenominator = priceDenominator
+  } else if (decimalsBase > decimalsQuote) {
+    newNumerator = priceNumerator.times(TEN.pow(u8(decimalsBase - decimalsQuote)))
```

Comment:
numerator/base/buy precision is > than denominator/quote/sell
In this case you should reduce the precision on numerator to match lower token precision, so you should divide instead of multiply:

```suggestion
    newNumerator = priceNumerator.div(TEN.pow(u8(decimalsBase - decimalsQuote)))
```

or, upscale denominator to match same precision as numerator.

Reply:
I think the mistake was other. There was a big confusing cause the units are hard to understand. See this long thread if you don't believe me:  https://gnosisinc.slack.com/archives/CPZA1AGKY/p1585149928002600